### PR TITLE
[FIX] KNN: Fix crash when Mahanalobis metric is used

### PR DESCRIPTION
**Pull request by VesnaT**

### CODE REVIEW

File: Orange/tests/test_knn.py [hunk 1]
```diff
@@ -7,23 +7,27 @@
...
+from Orange.regression import KNNRegressionLearner
 from Orange.evaluation import CA, CrossValidation
 
 
 class TestKNNLearner(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
         cls.iris = Table('iris')
-        cls.learn = KNNLearner()
+        cls.learn = KNNLearner
```
Comment (kernc) [inline Orange/tests/test_knn.py]: What's wrong with a single learner instance?

File: Orange/classification/knn.py [hunk 2]
```diff
@@ -11,5 +11,7 @@ class KNNLearner(SklLearner):
     def __init__(self, n_neighbors=5, metric="euclidean", weights="uniform",
                  algorithm='auto',
                  preprocessors=None):
+        if metric == "mahalanobis":
+            algorithm = "brute"
```
Comment (kernc) [inline Orange/classification/knn.py]: Please add a comment explaining why this is necessary. For the record, this works for me:
```py
>>> from sklearn.neighbors import KNeighborsClassifier
>>> nn = KNeighborsClassifier(metric='mahalanobis', algorithm='auto')
>>> X = np.random.random((10, 2))
>>> y = (np.random.random(10) > .5).astype(int)
>>> nn.fit(X, y)
>>> nn.predict(X)
array([1, 1, 1, 1, 0, 0, 0, 0, 0, 0])
```

Comment (VesnaT) [inline Orange/classification/knn.py]: For me as well. But i.e. X = np.random.random((12, 2)) doesn't.

File: Orange/tests/test_knn.py [hunk 3]
```diff
@@ -7,23 +7,27 @@
...
+from Orange.regression import KNNRegressionLearner
+from Orange.evaluation import CA, CrossValidation, MSE
 
 
 class TestKNNLearner(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
         cls.iris = Table('iris')
-        cls.learn = KNNLearner()
+        cls.learn = KNNLearner
```
Comment (astaric) [inline Orange/tests/test_knn.py]: I would skip assigning classes in to cls.learn and cls.learn_reg and just use their names in tests.

self.learn() is not much shorter than KNNLearner() (in fact it has the same number of characters)

File: Orange/widgets/classify/tests/test_owknnclassification.py [hunk 4]
```diff
@@ -1,53 +1,30 @@
...
-from PyQt4 import QtGui
-
-from Orange.widgets.tests.base import WidgetTest
 from Orange.widgets.classify.owknn import OWKNNLearner
+from Orange.widgets.tests.base import (WidgetTest, WidgetLearnerTestMixin,
+                                       GuiToParam)
 
 
-class TestOwKnnClassification(WidgetTest):
+class TestOWKNN(WidgetTest, WidgetLearnerTestMixin):
```
Comment (astaric) [inline Orange/widgets/classify/tests/test_owknnclassification.py]: If the test case name matches TestClassName syntax, PyCharm will allow you to jump between tests and code using a keyboard shortcut (CMD + T on mac)

File: Orange/widgets/classify/tests/test_owknnclassification.py [hunk 5]
```diff
@@ -1,53 +1,30 @@
...
-    def test_learner_on_output(self):
-        """ Check if learner is on output after create widget  and apply """
-        self.widget.apply()
-        self.assertNotEqual(self.widget.learner, None)
-
+        self.widget = self.create_widget(OWKNNLearner,
+                                         stored_settings={"auto_apply": False})
+        self.init()
+
+        def combo_set_value(i, x):
```
Comment (astaric) [inline Orange/widgets/classify/tests/test_owknnclassification.py]: I have seen this in multiple test classes, should we change it to a static method on WidgetLearnerTestMixin?

Comment (VesnaT) [inline Orange/widgets/classify/tests/test_owknnclassification.py]: Yes, we should. Was planning to do it after it is merged, to avoid conflicts.

Comment (astaric) [inline Orange/widgets/classify/tests/test_owknnclassification.py]: ok
